File: 程序员面试金典/5.8绘制直线/drawLine.java
class Solution {
    public int[] drawLine(int length, int w, int x1, int x2, int y) {
        int n = w / 32;
        int h = length / n;
        int[] screen = new int[length];
        for (int i = x1; i <= x2; ++i)
            screen[y * n + i / 32] |= (1 << (31 - i % 32));
        return screen;
    }
}

File: 程序员面试金典/5.8/drawLine.py
class Solution:
    def drawLine(self, length: int, w: int, x1: int, x2: int, y: int) -> List[int]:
        n = w // 32
        h = length // n
        screen = [0] * length
        c1, b1 = divmod(x1, 32)
        c2, b2 = divmod(x2, 32)
        if c1 == c2:
            screen[y * n + c1] |= ((1 << (b2 - b1 + 1)) - 1) << (31 - b2)
        else:
            screen[y * n + c1] |= (1 << (32 - b1)) - 1
            for c in range(c1 + 1, c2):
                screen[y * n + c] = -1
            screen[y * n + c2] |= -1 << (31 - b2)
        return screen
